python/rainbow/simulators/prox_soft_bodies/gauss_seidel_parallel.py

```diff
@@ -0,0 +1,381 @@
+import numpy as np
+import numba as nb
+import networkx as nx
+from collections import defaultdict
+from rainbow.util.timer import Timer
+
+
+def _build_contact_graph(J):
+    """ Create the contact graph based on the contact jacobi matrix.
+        The vertices of the graph are the contact points, and the edges of the graph are the value of the contact jacobi matrix is not zero between two contact points.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        graph: The contact graph.
+    """
+    G = nx.Graph()
+    K = J.shape[0] // 4
+    G.add_nodes_from(np.arange(K))
+    sources, targets = _get_edges(J)
+    for s, t in zip(sources, targets):
+        G.add_edge(s, t) 
+    
+    return G
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def _get_edges(J):
+    """ Get the edges of the contact graph.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        (ArrayLike, ArrayLike): The source and target of the edges.
+    """
+    K = J.shape[0] // 4
+    step_size = 24
+    cols = J.shape[1]
+    
+    # Preallocate arrays for source and target of edges
+    max_possible_edges = (cols // step_size) * K * K  
+    source = np.full(max_possible_edges, -1, dtype=np.int64)
+    target = np.full(max_possible_edges, -1, dtype=np.int64)
+    edge_count = 0
+
+    for idx in nb.prange(0, cols//step_size):
+        col_idx = idx * step_size
+        contacts = np.where(J[:, col_idx: col_idx+step_size])[0]
+        contacts = np.unique(contacts // 4)
+        for i, contact_i in enumerate(contacts):
+            for j, contact_j in enumerate(contacts):
+                if i < j:
+                    source[edge_count] = contact_i
+                    target[edge_count] = contact_j
+                    edge_count += 1
+
+    # Trim the arrays to the actual size
+    valid_indices = source != -1
+    filtered_source = source[valid_indices]
+    filtered_source = target[valid_indices]
```

Comment:
Is this really the most efficient way to extract the edge information?

scipy has an inbuilt function that should give you almost what you need

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.nonzero.html

You would just need to do modulo of indices to get "block" indices ?

Here is a suggestion

`def find_filled_block_indices(sparse_matrix, block_size=(4, 3)):
    # Extracting row and column indices of non-zero elements
    rows, cols = sparse_matrix.nonzero()

    # Calculate block indices
    block_row_indices = rows // block_size[0]
    block_col_indices = cols // block_size[1]

    # Combine and remove duplicates
    block_indices = set(zip(block_row_indices, block_col_indices))

    return block_indices
`

Reply:
Numba can not support sparse matrix. Therefore, those functions can not be used in a Numba function.
But we could use the scipy instead of Numba in here. I will test which one is better to get higher performance.